addons/coupon/models/coupon_program.py

```diff
@@ -93,12 +93,15 @@ def create(self, vals):
 
     def write(self, vals):
         res = super(CouponProgram, self).write(vals)
+        if not self:
+            return res
         reward_fields = [
             'reward_type', 'reward_product_id', 'discount_type', 'discount_percentage',
             'discount_apply_on', 'discount_specific_product_ids', 'discount_fixed_amount'
         ]
         if any(field in reward_fields for field in vals):
-            self.mapped('discount_line_product_id').write({'name': self[0].reward_id.display_name})
+            for reward_id in self.reward_id:
+                self.search([('reward_id', '=', reward_id.id)]).mapped('discount_line_product_id').write({'name': reward_id.display_name})
```

Comment:
```suggestion
            for program in self:
                program.discount_line_product_id.write({'name': self.reward_id.display_name})
```
`search` calls make one database request by call, when we already have the information in cache to treat it without search calls.
IMO, we won't gain much performance (or even lose some) by trying to group the writes here, let's use a simple loop for now ;).